### PR TITLE
Add a type field to struct Event

### LINKED ISSUE #72: Add a type field to struct Event
Hi would like to request if its possible to add a `type` property to struct `Event` here https://github.com/iotaledger/wallet.rs/blob/bbebd26198732e768cf58ff407fbe3cdcf01c5e4/src/events/types.rs#L14

Currently the type of the event appears as the Key of the json instead of the Value.

Eg.

```json
{
  "accountIndex": 0,
  "event": {
    "SpentOutput": { # Appearing as key
      "output": {
        "outputId": "0xe23d69a6fc1e859ab3f778f907cf8bf31517dea65fb6577396101f6ffd5c6ef90000",
        "metadata": {
          "blockId": "0x1798b047a139a6679de33e97e86dfedbdfd3554b41be66927b50b68d7747f043",
          "transactionId": "0xe23d69a6fc1e859ab3f778f907cf8bf31517dea65fb6577396101f6ffd5c6ef9",
          "outputIndex": 0,
          "isSpent": true,
          "milestoneIndexSpent": 2492220,
          "milestoneTimestampSpent": 1670597393,
          "transactionIdSpent": "0x2d8dd8a23e86e1438efcd236ccfffa75a45c1d5807c05f1ec60dece7e12ae830",
          "milestoneIndexBooked": 2492178,
          "milestoneTimestampBooked": 1670597183,
          "ledgerIndex": 2518863
        },
        "output": {
          "type": 3,
          "amount": "1000000000",
          "unlockConditions": [
            {
              "type": 0,
              "address": {
                "type": 0,
                "pubKeyHash": "0x03b75085d5fab12547c9b20a4258d3bf47c07c0d37ea030dceb44cd63bf09ea4"
              }
            }
          ]
        },
        "isSpent": true,
        "address": {
          "type": 0,
          "pubKeyHash": "0x03b75085d5fab12547c9b20a4258d3bf47c07c0d37ea030dceb44cd63bf09ea4"
        },
        "networkId": "8342982141227064571",
        "remainder": false,
        "chain": [
          {
            "hardened": true,
            "bs": [
              128,
              0,
              0,
              44
            ]
          },
          {
            "hardened": true,
            "bs": [
              128,
              0,
              16,
              123
            ]
          },
          {
            "hardened": true,
            "bs": [
              128,
              0,
              0,
              0
            ]
          },
          {
            "hardened": true,
            "bs": [
              128,
              0,
              0,
              0
            ]
          },
          {
            "hardened": true,
            "bs": [
              128,
              0,
              0,
              1
            ]
          }
        ]
      }
    }
  }
}
```

I find this would make it easier to serialize it and wont impose any breaking changes to the current bindings.

This would also make it uniform to `iota.js` which uses a variable `type` to distinguish between `AddressTypes` like `Alias`,`Ed25519` or `Nft`.

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes #72

### CODE REVIEW

**Review by kwek20 — COMMENTED** [review]
needs a changelog but otherwise lgtm

**Review by Thoralf-M — APPROVED** [review]

**Review by kwek20 — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

**Comment by thibault-martinez** [comment]
Don't merge yet pls

File: sdk/tests/wallet/events.rs
Comment (Thoralf-M) [inline sdk/tests/wallet/events.rs:26]: Nice one

File: sdk/src/wallet/events/types.rs
Comment (Thoralf-M) [inline sdk/src/wallet/events/types.rs:40]: Changelog and nodejs types also need to be updated?

File: sdk/src/wallet/events/types.rs
Comment (Alex6323) [inline sdk/src/wallet/events/types.rs:28]: putting this comment here ... but can you add a blank line after line 17? Why is the formatter not adding one? I though it would do that.

Comment (thibault-martinez) [reply sdk/src/wallet/events/types.rs:28]: Good catch, and no, formatter doesn't do that kind of changes unfortunately

**Review by Alex6323 — DISMISSED** [review]

**Review by kwek20 — DISMISSED** [review]

File: bindings/nodejs/lib/types/wallet/event.ts
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/wallet/event.ts:10]: Do we really not have this type anywhere?

File: bindings/wasm/src/wallet.rs
Comment (thibault-martinez) [inline bindings/wasm/src/wallet.rs]: I am 0% sure about that, but I have no clue what to do about it

File: sdk/src/wallet/bindings/nodejs/types/event.ts
Comment (thibault-martinez) [inline sdk/src/wallet/bindings/nodejs/types/event.ts:13]: ????

File: sdk/src/wallet/events/types.rs
Comment (thibault-martinez) [reply sdk/src/wallet/events/types.rs:40]: Done

File: bindings/nodejs/lib/types/wallet/event.ts
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/wallet/event.ts:10]: I don't think so, in iota.js the parameters are also `transactionId: string`

File: bindings/nodejs/lib/types/wallet/event.ts
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/wallet/event.ts:40]: Currently we never emit this event :trollface:

File: bindings/nodejs/lib/types/wallet/event.ts
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/wallet/event.ts]:
```suggestion
    * The transaction inclusion state
```

File: bindings/wasm/src/wallet.rs
Comment (Thoralf-M) [reply bindings/wasm/src/wallet.rs]: Why did you change it?

File: sdk/src/wallet/bindings/nodejs/types/event.ts
Comment (Thoralf-M) [inline sdk/src/wallet/bindings/nodejs/types/event.ts]:
```suggestion
    * The transaction inclusion state.
```

File: bindings/wasm/src/wallet.rs
Comment (thibault-martinez) [reply bindings/wasm/src/wallet.rs]: Because it's not an array of string now, it's an array of integers and I have absolutely no clue how to get them out

Comment (thibault-martinez) [reply bindings/wasm/src/wallet.rs]: Can't even find it online 🤷🏻‍♂️

Comment (Thoralf-M) [reply bindings/wasm/src/wallet.rs]: I think `event_type.unwrap().as_f64().unwrap() as u8` should work, conversion of numbers from to JS is usually done via f64 I believe, but since it's just u8 your change should be fine as well

Comment (thibault-martinez) [reply bindings/wasm/src/wallet.rs]: Yeah I thought about that but it seemed a bit weird, I'll do it

Comment (thibault-martinez) [reply bindings/wasm/src/wallet.rs]: Like, why do they have as_f64 but not anything for actually integer?

File: sdk/src/wallet/bindings/nodejs/types/event.ts
Comment (thibault-martinez) [reply sdk/src/wallet/bindings/nodejs/types/event.ts]: Done

File: bindings/nodejs/lib/types/wallet/event.ts
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts]: Done

File: bindings/wasm/src/wallet.rs
Comment (thibault-martinez) [reply bindings/wasm/src/wallet.rs]: Done

File: bindings/nodejs/lib/types/wallet/event.ts
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts:10]: Removed the TODO

**Review by Thoralf-M — CHANGES_REQUESTED** [review]
Nodejs format

File: bindings/nodejs/examples/how_tos/events.ts
Comment (Thoralf-M) [inline bindings/nodejs/examples/how_tos/events.ts]:
```suggestion
// This example listens to wallet events.
```

File: bindings/nodejs/examples/how_tos/events.ts
Comment (Thoralf-M) [inline bindings/nodejs/examples/how_tos/events.ts:1]: Should this be in the how_tos?

Comment (thibault-martinez) [reply bindings/nodejs/examples/how_tos/events.ts:1]: I'll move it to examples

Comment (thibault-martinez) [reply bindings/nodejs/examples/how_tos/events.ts:1]: Done

File: bindings/nodejs/lib/types/wallet/event.ts
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/wallet/event.ts:48]: Shouldn't this also have the `accountIndex` field or would this be a new class?

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: Do we even need the `Event` with account_index? It seems to be unused? And always have been? Or am I missing something?

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: What's printed in the example is `{ accountIndex: 0, event: { type: 0 } }` so you know from/for which account this event is emitted

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: Ok but this was never done in bindings, right?

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: I can't find where I would have removed this

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: Yeah maybe it was never done
But what you get in the callback from listen() is something like
```TS
class Event {
    private accountIndex: number;
    private event: WalletEvent;

    constructor(accountIndex: number, event: WalletEvent) {
        this.accountIndex = accountIndex;
        this.event = event;
    }

    /**
     * The account index for which the event was emitted.
     */
    getAccountIndex(): number {
        return this.accountIndex;
    }

    /**
     * The wallet event.
     */
    getEvent(): WalletEvent {
        return this.event;
    }
}
```

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: Yeah, I'm doing this

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: With that it's possible to do 
```TS
        const callback = function (err: any, data: string) {
            const parsed = JSON.parse(data);
            const event = new Event(parsed.accountIndex, parsed.event)
            console.log('AccountIndex:', event.getAccountIndex());
            console.log('Event:', event.getEvent());
        };
```
and then get
```
AccountIndex: 0
Event: { type: 0 }
AccountIndex: 0
Event: { type: 5, progress: { type: 0 } }
```

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: Can do even better than that

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: Yeah it's ugly, please do even better :trollface:

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: Let me know what you think and I'll replicate to the other binding

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts:48]: I've done it to the old bindings as well

File: bindings/nodejs/examples/events.ts
Comment (kwek20) [inline bindings/nodejs/examples/events.ts:1]: move to wallet folder and use the ` const manager = await getUnlockedManager();`

File: sdk/src/wallet/bindings/java/lib/native/src/lib.rs
Comment (DaughterOfMars) [inline sdk/src/wallet/bindings/java/lib/native/src/lib.rs:206]: Can't use an array here?

Comment (thibault-martinez) [reply sdk/src/wallet/bindings/java/lib/native/src/lib.rs:206]: Sir this is Java, I just made it compile

Comment (thibault-martinez) [reply sdk/src/wallet/bindings/java/lib/native/src/lib.rs:206]: Let me see

**Review by kwek20 — COMMENTED** [review]
jsut the one example, and i think we can ignore the one in sdk/src/

File: sdk/src/wallet/bindings/java/lib/native/src/lib.rs
Comment (kwek20) [reply sdk/src/wallet/bindings/java/lib/native/src/lib.rs:206]: hes improving java, what a time to be alive!

Comment (thibault-martinez) [reply sdk/src/wallet/bindings/java/lib/native/src/lib.rs:206]: But count is not const?

Comment (DaughterOfMars) [reply sdk/src/wallet/bindings/java/lib/native/src/lib.rs:206]: Ah, bummer.

File: bindings/nodejs/examples/events.ts
Comment (thibault-martinez) [reply bindings/nodejs/examples/events.ts:1]: Eww when did we add that? We've renamed account manager to wallet in the new bindings, we need to change that ASAP

**Review by DaughterOfMars — DISMISSED** [review]

**Review by Alex6323 — DISMISSED** [review]

File: bindings/nodejs/examples/events.ts
Comment (thibault-martinez) [reply bindings/nodejs/examples/events.ts:1]: Done, I'll open an issue to rename all that stuff

**Review by DaughterOfMars — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Pull request merged**
